packages/mobx-state-tree/__tests__/core/model.test.ts

```diff
@@ -17,3 +17,152 @@ test("it should call preProcessSnapshot with the correct argument", () => {
+        })
+    })
+})
+describe("Model properties", () => {
+    test("Providing a string as the first argument and an object as the second argument should use the object's properties in the model.", () => {
+        const Model = types.model("name", {
+            prop1: "prop1",
+            prop2: 2
+        })
+
+        expect(Model.properties).toHaveProperty("prop1")
+        expect(Model.properties).toHaveProperty("prop2")
+    })
+    test("Providing an object as the first argument should parse and use its properties.", () => {
+        const Model = types.model({
+            prop1: "prop1",
+            prop2: 2
+        })
+
+        expect(Model.properties).toHaveProperty("prop1")
+        expect(Model.properties).toHaveProperty("prop2")
+    })
+    test("Providing a string as the first argument and a falsy value as the second argument should result in an empty set of properties.", () => {
+        const Model = types.model("name", null as any)
+
+        expect(Model.properties).toEqual({})
+    })
+})
+describe("Model identifier", () => {
+    test("If no identifier attribute is provided, the identifierAttribute should be undefined.", () => {
+        const Model = types.model("name", {})
+
+        expect(Model.identifierAttribute).toBeUndefined()
+    })
+    test("If an identifier attribute is provided, the identifierAttribute should be set for the object.", () => {
+        const Model = types.model("name", {
+            id: types.identifier
+        })
+
+        expect(Model.identifierAttribute).toBe("id")
+    })
+    test("If an identifier attribute has already been provided, an error should be thrown when attempting to provide a second one.", () => {
+        expect(() => {
+            types.model("name", {
+                id: types.identifier,
+                id2: types.identifier
+            })
+        }).toThrowErrorMatchingInlineSnapshot(
+            `"[mobx-state-tree] Cannot define property 'id2' as object identifier, property 'id' is already defined as identifier property"`
+        )
+    })
+})
+describe("Edge case behavior", () => {
+    describe("when we provide no arguments to the function", () => {
+        test("the model will be named AnonymousModel", () => {
+            const Model = types.model()
+
+            expect(Model.name).toBe("AnonymousModel")
+        })
+        test("the model will have no properties", () => {
+            const Model = types.model()
+
+            const modelSnapshot = getSnapshot(Model.create())
+            expect(modelSnapshot).toEqual({})
+        })
+    })
+    describe("when we provide an invalid name value, but a valid property object", () => {
+        test("the model will be named AnonymousModel", () => {
+            const Model = types.model(null as any, {
+                prop1: "prop1",
+                prop2: 2
+            })
+
+            expect(Model.name).toBe("AnonymousModel")
+        })
+        test("the model will have no properties", () => {
```

Comment:
@jamonholmgren - to me, this is possibly unintended behavior? It's probably not a huge deal. For folks using TypeScript, they'll get a type error if they try to:

```ts
const Model = types.Model(null, { prop1: "p1"}) // Redline on the first argument.
```

But for JavaScript consumer, this same code won't give them any TS hints, nor will it error at runtime. But the properties will end up empty.

I think this is a reasonable failure mode - nice that it won't actually toss an exception, but maybe confusing, since we don't even log anything in dev mode.

Reply:
I think we should throw an error for sure; unintended behavior. I see this as a bugfix and not a breaking change.

Where this could come into play is if you end up constructing the model name using a variable and accidentally not getting anything.

```js
let modelName = undefined
if (someCondition) {
  modelName = "HelloWorld"
}

const Model = types.Model(modelName, { prop1: "p1"})
```

If `someCond` is truthy, it'll work as expected. If it's falsy, you'll just get really weird, confusing problems.

Reply:
Yeah I agree. I'm going to file this as an issue. Should be an easy fix - I'll tag it as such and save it to see if any new contributors want to pick it up sometime since it's not urgent.

Reply:
Opened an issue: https://github.com/mobxjs/mobx-state-tree/issues/2045